feat(System): introduce new switch leader button

vrozaev opened the pull request:
<!-- nda-start -->
https://nda.ya.ru/t/EnPm76mE75u5a4
<!-- nda-end -->In this PR I introduce a new button and dialog which switch leader for primary masters.
 
You can find this new button on the /system page, in masters section:
<img width="585" alt="330418244-d1a5c91a-e719-4ba7-9515-d27f8e1d3316" src="https://github.com/ytsaurus/ytsaurus-ui/assets/2428161/9107b0d2-07eb-4bec-aa89-d3a5362e153f">

The feature was developed and tested only for primary masters because I don't have a cluster with secondary masters. 

PS: that my first meaningful PR in this project, so please give as much feedback as possible, I need to know how to develop things in right way.

ma-efremoff commented on packages/ui/src/ui/pages/system/Masters/SwitchLeader.tsx:
```diff
@@ -23,42 +23,61 @@ type SwitchLeaderPromtProps = {
             visible={props.visible}
-            loading={props.loading}
+            headerProps={{
+                title: 'Are you sure that you want to switch the cell leader?',
```
Switch leader

ma-efremoff replied:
title: 'Switch leader'

ma-efremoff commented on packages/ui/src/ui/pages/system/Masters/SwitchLeader.tsx:
```diff
@@ -23,42 +23,61 @@ type SwitchLeaderPromtProps = {
+            headerProps={{
+                title: 'Are you sure that you want to switch the cell leader?',
+            }}
+            fields={[
```
```ts
initialValues={{
    old_leader: props.currentLeaderAddress
    new_leader: props.newLeaderAddress
}}
fields={[{
    type: 'plain',
    caption: 'Old leader',
    name: 'old_leader',
}, {
    type: 'plain'
    caption: 'New leader',
    name: 'new_leader'
}, {
    type: 'block',
    name: 'text',
    extras: ({new_leader}) => {
        return {
             children: <span>{`Are you sure you want to switch leader cell to `${new_leader}`?}</span>;
        }
    }
}, 
...makeErrorFields([error])
]}
```

ma-efremoff commented on packages/ui/src/ui/pages/system/Masters/SwitchLeader.tsx:
```diff
@@ -23,42 +23,61 @@ type SwitchLeaderPromtProps = {
+                },
+            ]}
+            footerProps={{
+                textApply: 'Switch',
```
textApply: 'Switch leader'

ma-efremoff commented on packages/ui/src/ui/pages/system/Masters/SwitchLeader.tsx:
```diff
@@ -23,42 +23,61 @@ type SwitchLeaderPromtProps = {
+            footerProps={{
+                textApply: 'Switch',
+            }}
+            onAdd={(_form) => {
```
```ts
onAdd={(form) => {
     const {new_leader} = form.getState().values;
     return props.confirm(new_leader)
        .then(() => {
              setError(undefined);
         })
         .catch((e) => {
             setError(e);
             throw e;
         });
}}
```

ma-efremoff commented on packages/ui/src/ui/pages/system/Masters/SwitchLeader.tsx:
```diff
@@ -23,42 +23,61 @@ type SwitchLeaderPromtProps = {
+                return Promise.resolve();
+            }}
+            onClose={props.cancel}
+            isApplyDisabled={isApplyDisabled}
```
Use `pristineSubmittable={true}` instead of `isApplyDisabled`. isApplyDisabled is required when you need to check values of form.

ma-efremoff commented on packages/ui/src/ui/pages/system/Masters/SwitchLeader.tsx:
```diff
@@ -0,0 +1,145 @@
+    const isApplyDisabled = useCallback(() => false, []);
+
+    return (
+        <YTDFDialog
```
```ts
// Somewhere above
type FormValues = {
    old_leader: string;
    new_leader: string;
}

// here
return (
    <YTDFDialog<FormValues>
        ...
    />
);

```

ma-efremoff commented on packages/ui/src/ui/pages/system/Masters/SwitchLeader.tsx:
```diff
@@ -0,0 +1,145 @@
+
+const SwitchLeaderPromt = (props: SwitchLeaderPromtProps) => {
+    const isApplyDisabled = useCallback(() => false, []);
+
```
It is better to add error state:
```ts
const [error, setError] = React.useState<YTError | undefined>(undefined);
```

ma-efremoff commented on packages/ui/src/ui/pages/components/SwitchLeaderShortInfo/SwitchLeaderShortInfo.scss:
```diff
@@ -0,0 +1,2 @@
+.switch-leader-short-info {
```
looks like an unnecessary file

ma-efremoff commented on packages/ui/src/ui/pages/system/Masters/SwitchLeader.tsx:
```diff
@@ -0,0 +1,145 @@
+    newLeaderAddress: string;
+};
+
+const SwitchLeaderPromt = (props: SwitchLeaderPromtProps) => {
```
We prefer `Dialog`/`Modal` suffix for such components. It is not an issue, but can you please rename it to `SwitchLeaderDialog`/`SwitchLeaderModal`.

ma-efremoff commented:
- `Switch leader` dialog should display current leader
- `switch_leader` command doesn't work when UI displays `Host`, but works for `Containers` mode
- `//sys/primary_masters/{new_leader_address}/orchid/monitoring/hydra` waits until new leader is activated, i.e. it is called only once, but we expect polling requests every 3 seconds (we need to ask someone from API-team if it is possible to avoid it)

vrozaev commented:
> Switch leader dialog should display current leader
Fixed here: https://github.com/ytsaurus/ytsaurus-ui/pull/503/commits/51eec9ebbdf1ba08aea0d8beeb073744d8baf6dc

> switch_leader command doesn't work when UI displays Host, but works for Container mode
Fixed here: https://github.com/ytsaurus/ytsaurus-ui/pull/503/commits/350bc17b6bcec9b070bfe9b3db3c69a096532811
